picard/util/checkupdate.py

```diff
@@ -0,0 +1,181 @@
+        '''Provide a list of the latest version tuples for each update type.'''
+        return self._available_versions
+
+    def _query_available_updates(self, callback=None):
+        '''Gets list of releases from specified website api.'''
+        log.debug(_("Getting release information from {host_url}.".format(host_url=VERSIONS_API['host'],)))
+        self.tagger.webservice.get(
+            VERSIONS_API['host'],
+            VERSIONS_API['port'],
+            VERSIONS_API['endpoint'],
+            partial(self._releases_json_loaded, callback=callback),
+            parse_response_type=None,
+            priority=True,
+            important=True
+        )
+
+    def _releases_json_loaded(self, response, reply, error, callback=None):
+        '''Processes response from specified website api query.'''
+        if error:
+            log.error(_("Error loading releases list: {error_message}").format(error_message=reply.errorString(),))
+            if self._show_always:
+                QMessageBox.information(
+                    self._parent,
+                    _("Picard Update"),
+                    _("Unable to retrieve the latest version information."),
+                    QMessageBox.Ok, QMessageBox.Ok)
+        else:
+            self._available_versions = load_json(response)['versions']
+            for key in self._available_versions.keys():
+                log.debug("Version key '{version_key}' --> {version_information}".format(
+                    version_key=key, version_information=self._available_versions[key],))
+            self._display_results()
+        if callback:
+            callback(not error)
+
+    def _display_results(self):
+        # Display results to user.
+        key = ''
+        high_version = PICARD_VERSION
+        for test_key in PROGRAM_UPDATE_LEVELS:
+            if self._update_level >= PROGRAM_UPDATE_LEVELS[test_key]['level'] and  compare_version_tuples(high_version, self._available_versions[test_key]['version']) > 0:
+                key = test_key
+                high_version = self._available_versions[test_key]['version']
+        if key:
+            if QMessageBox.information(
+                self._parent,
+                _("Picard Update"),
+                _("A new version of Picard is available.\n\n"
+                  "Old version: {picard_old_version}\n"
+                  "New version: {picard_new_version}\n\n"
+                  "Would you like to download the new version?").format(
+                      picard_old_version=PICARD_FANCY_VERSION_STR,
+                      picard_new_version=self._available_versions[key]['tag']
+                    ),
+                QMessageBox.Ok | QMessageBox.Cancel,
+                QMessageBox.Cancel
+            ) == QMessageBox.Ok:
+                wb2.open(self._available_versions[key]['urls']['download'])
+        else:
+            if self._show_always:
+                for key in PROGRAM_UPDATE_LEVELS.keys():
```

Comment:
`.keys()` not needed here.
But you could write:

```python
for key, update in PROGRAM_UPDATE_LEVELS.items():
    if self._update_level == update['level']:
        update_level_text = update['title']
```

**But i have the feeling you could simplify code even more if you modify `PROGRAM_UPDATE_LEVELS` to have numeric level as key:**

```python
PROGRAM_UPDATE_LEVELS = {
    0: {
        'name': 'stable',
        'title': N_('Stable releases only'),
    },
   ...
```

then this loop would be uneeded:

```python
update_level_text = PROGRAM_UPDATE_LEVELS[self._update_level]['title']
```

Reply:
I was toying with having numeric keys originally, but don't recall why I ended up not doing it.  I'll look at it again because it seems to make much more sense with the numeric keys.